packages/block-events-stream/src/createBlockEventsStream.ts

```diff
@@ -0,0 +1,126 @@
+import { BlockEvents, BlockEventsStream, ReadonlyBehaviorSubject } from "./common";
+import { bigIntMin } from "./utils";
+import { isNonPendingLog } from "./isNonPendingLog";
+import { debug } from "./debug";
+import { createBlockNumberStream } from "./createBlockNumberStream";
+import { getLogs } from "./getLogs";
+import { storeEventsAbi } from "@latticexyz/store";
+
+export type CreateBlockEventsStreamOptions<TAbiEvent extends AbiEvent> = {
+  publicClient: PublicClient;
+  fromBlock?: BlockNumber;
+  toBlock?: BlockNumber | ReadonlyBehaviorSubject<BlockNumber>;
+  address?: Hex;
+  events: readonly TAbiEvent[];
+  maxBlockRange?: number; // defaults to 1000
+};
+
+export async function createBlockEventsStream<TAbiEvent extends AbiEvent>({
+  publicClient,
+  fromBlock: initialFromBlock,
+  toBlock: initialToBlock,
+  address,
+  events,
+  maxBlockRange = 1000,
+}: CreateBlockEventsStreamOptions<TAbiEvent>): Promise<BlockEventsStream<TAbiEvent>> {
+  debug("createBlockEventsStream", { initialFromBlock, initialToBlock, address, events, maxBlockRange });
+
+  if (initialFromBlock == null) {
+    debug("getting earliest block");
+    const earliestBlock = await publicClient.getBlock({ blockTag: "earliest" });
+    debug("earliest block", earliestBlock);
+    if (earliestBlock.number == null) {
+      // TODO: better error
+      throw new Error(`pending or missing earliest block`);
+    }
+    initialFromBlock = earliestBlock.number;
+  }
+
+  if (initialToBlock == null) {
+    debug("creating latest block number stream");
+    initialToBlock = await createBlockNumberStream({ publicClient, blockTag: "latest" });
+  }
+
+  const stream = new Subject<BlockEvents<TAbiEvent>>();
+  fetchBlockRange(
+    initialFromBlock,
+    maxBlockRange,
+    initialToBlock instanceof BehaviorSubject ? initialToBlock.value : initialToBlock
+  );
+
+  async function fetchBlockRange(fromBlock: bigint, maxBlockRange: number, lastBlockNumber: bigint): Promise<void> {
+    try {
+      const toBlock = bigIntMin(fromBlock + BigInt(maxBlockRange), lastBlockNumber);
+      debug("fetching block range", { fromBlock, toBlock });
+
+      // TODO: swap this with viem `getLogs` call when viem supports multiple events: https://github.com/wagmi-dev/viem/pull/633
+      const logs = await getLogs({
+        publicClient,
+        address,
+        fromBlock,
+        toBlock,
+        events: storeEventsAbi,
+      });
+
+      // TODO: do something other than just throwing out pending logs
+      const nonPendingLogs = logs.filter(isNonPendingLog);
+
+      if (logs.length !== nonPendingLogs.length) {
+        // TODO: better error
+        console.warn("pending logs discarded");
+      }
+
+      // TODO: handle RPC block range errors
+      // TODO: handle RPC rate limit errors (hopefully via client retry policy)
+
+      const blockNumbers = Array.from(new Set(nonPendingLogs.map((log) => log.blockNumber)));
+      blockNumbers.sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
```

Comment:
I think this was leftover after some refactoring where `a` and `b` were previously something like `a.blockNumber` and `b.blockNumber` - will simplify!

Reply:
oh I remember why: it's because these are bigints and native `.sort` doesn't handle bigints 🤦 

<img width="488" alt="image" src="https://github.com/latticexyz/mud/assets/508855/35588f28-10c8-42f4-b48c-d59a37e8b76e">